apiserver/pkg/http/client.go

```diff
@@ -638,29 +650,90 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
 }
 
 func (krc *KuberayAPIServerClient) executeRequest(httpRequest *http.Request, URL string) ([]byte, *rpcStatus.Status, error) {
-	response, err := krc.httpClient.Do(httpRequest)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to execute http request for url '%s': %w", URL, err)
-	}
-	defer func() {
-		if closeErr := response.Body.Close(); closeErr != nil {
-			klog.Errorf("Failed to close http response body because %+v", closeErr)
+	// Set the overall timeout
+	ctx, cancel := context.WithTimeout(context.Background(), krc.retryCfg.OverallTimeout)
+	defer cancel()
+
+	httpRequest = httpRequest.WithContext(ctx)
+
+	// Record the last error and status got
+	var lastErr error
+	var lastStatus *rpcStatus.Status
+
+	// Only retry for HTTP status codes defined as retryable in isRetryableHTTPStatus().
+	for attempt := 0; attempt <= krc.retryCfg.MaxRetry; attempt++ {
+		response, err := krc.httpClient.Do(httpRequest)
+		// Error in sending the request, treated as non-retryable error
+		if err != nil {
+			lastStatus = nil
+			lastErr = fmt.Errorf("failed to execute http request for url '%s': %w", URL, err)
+			break
 		}
-	}()
-	bodyBytes, err := io.ReadAll(response.Body)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to read response body bytes: %w", err)
-	}
-	if response.StatusCode != http.StatusOK {
+
+		defer func() {
+			if closeErr := response.Body.Close(); closeErr != nil {
+				klog.Errorf("failed to close http response body because %+v", closeErr)
+			}
+		}()
```

Comment:
We can't use defer to close the per loop response.Body in this way. deferred blocks are only be executed at the end of `executeRequest`. Consider using this instead:

```go
doReq = func() (response, error) {
    ...
    defer func() {
        response.Body.Close()
    }()
    ....
}
for attempt := 0; attempt <= krc.retryCfg.MaxRetry; attempt++ {
    response, err := doReq(httpRequest)
}
```